README
Welcome

Find info about the library in the [wiki](https://github.com/architek/ccsds-standalone/wiki)


Ccsds/Changes
2010-07-18 13:35:26 +0200  [f7fc167] Bumped to version 1.7 
2010-07-18 13:33:30 +0200  [c3bd55c] Added TM Time case 
2010-07-18 13:32:32 +0200  [96294ab] Added TC Source Packet Printer 
2010-07-18 13:17:24 +0200  [4e39c2d] Added autoflush 
2010-07-18 13:07:17 +0200  [d6b7a92] Added FEE example 
2010-07-02 15:31:13 +0200  [ed36bc7] Offset comments 
2010-07-02 13:45:54 +0200  [78b347a] Updated changes file 
2010-07-02 13:40:27 +0200  [716cdcf] Removed old stuff and add version display to cltu.pl 
2010-07-02 13:36:23 +0200  [ac58dd5] Bump to 1.6 as decode_cltu_data API changed 
2010-07-02 13:36:04 +0200  [7df70be] Cltu Printer to visualize CBH bytes 
2010-07-02 13:34:38 +0200  [103f98c] Changed cltu decoder state machine and moved cbh handling in Utils 
2010-07-02 03:11:55 +0200  [4fd9369] Corrected MANIFEST 
2010-07-02 03:04:58 +0200  [c8aa578] Removed test file 
2010-07-02 03:00:18 +0200  [9b001b3] Update change file 
2010-07-02 02:25:39 +0200  [7ff2704] Correct debug to be exported 
2010-07-02 02:24:44 +0200  [bb3a440] Bump to version 1.5 
2010-07-02 01:51:02 +0200  [9f79304] Added Comments on size of structures 
2010-07-02 01:50:09 +0200  [bf9cd31] TM Frame Header defined 
2010-07-02 01:49:37 +0200  [4291512] Modified debug output 
2010-07-01 20:36:44 +0200  [eaa2a6b] Clean code 
2010-07-01 19:33:30 +0200  [670d482] Segmentation layer state machine added to example 
2010-07-01 17:27:52 +0200  [135665e] Example of decoding a cltu 
2010-07-01 17:22:17 +0200  [b48aa60] Added Cltu 
2010-06-30 02:01:25 +0200  [5f4dcfc] Added TC Frame 
2010-06-29 22:36:41 +0200  [ae35c73] Error left when package was put out of Data::ParseBinary 
2010-06-17 19:09:55 +0200  [0e6e907] Version in pod part removed in Utils.pm 
2010-06-17 17:49:26 +0200  [678bfb4] There was no error in pod. Revert 5fcb75121982aa7fc4024c0bda6504e9be4b9880 
2010-06-12 15:01:05 +0200  [5fcb751] Error in pod 
2010-06-12 14:57:33 +0200  [fd9ea7c] Man pages cleaned 
2010-06-12 12:56:48 +0200  [507a573] Put Ccsds out of perl standard Data::ParseBinary 
2010-06-12 12:42:18 +0200  [f214543] Adapted to new lib name 
2010-06-08 10:04:09 +0200  [6574013] Removed useless code from example 
2010-06-08 01:45:01 +0200  [8cca32e] Version changed to 1.3 
2010-06-08 01:43:07 +0200  [09aec4e] Obcp was not properly linked to TCs and few bugs corrected 
2010-06-07 09:23:21 +0200  [d6d3f31] Added proper Crc inline patching 
2010-06-03 13:50:05 +0200  [1b06c7a] New example 
2010-06-03 00:54:05 +0200  [5f9de9b] Commented useless code 
2010-06-03 00:50:35 +0200  [04b1765] 3 small examples 
2010-06-02 20:20:31 +0200  [18c746c] Bumped version to 1.2 
2010-06-02 19:07:26 +0200  [07592c1] Moved the 2 decoder examples to examples dir 
2010-06-02 19:06:25 +0200  [1faa198] Reorganized directory structure and removed bin for now 
2010-06-02 02:15:37 +0200  [212f1ea] Moved some code 
2010-06-02 02:08:54 +0200  [abb78fd] Bug in TMPrint for incl. TCs 
2010-06-02 01:21:06 +0200  [7c8278d] Reverted error introduced by a6c72fb6 
2010-05-14 11:59:40 +0200  [e5991ce] New binary includes perl packaged 
2010-05-14 11:02:44 +0200  [7b258cc] Added bin 
2010-05-12 08:48:15 +0200  [fc15587] Tidy, integrate TC to TM Pus, moved Apid to common 
2010-05-12 00:20:06 +0200  [2cbc4e2] TM PUS is now using TC Sourcepacket 
2010-05-11 22:39:16 +0200  [785e9c2] TC Handling minimal but should work 
2010-05-11 21:33:33 +0200  [27c1a1c] Changed version to 1.1 for TC handling 
2010-05-11 21:32:14 +0200  [9a19b44] Add small TC handling 
2010-05-11 19:26:56 +0200  [75570e3] Indent 
2010-05-11 08:56:09 +0200  [ddf81b5] Doc, removed some TODOs 
2010-05-11 08:29:54 +0200  [a6c72fb] Merge, identing, version bumped to 1.02 
2010-05-10 21:13:35 +0200  [3ab32fb] Tidy, crc, etc.. 
2010-04-27 01:09:35 +0200  [d0d1c04] bin crc 
2010-04-26 08:01:15 +0200  [f5e169d] META.yml regenerated 
2010-04-25 11:01:17 +0200  [45e72e7] Version 1.01 
2010-04-25 11:00:04 +0200  [de1db1c] Error in utils 
2010-04-25 01:21:04 +0200  [4ef8c53] Added requires 
2010-04-25 00:40:14 +0200  [8802c8e] switched to Module::Install 
2010-04-24 22:51:01 +0200  [2857c99] Lots of changes 
2010-04-21 01:08:42 +0200  [6e501be] Bumped to version 1.00 and updated decoders 
2010-04-20 23:22:48 +0200  [b88fb2b] Added back decoders using the module and test datas 
2010-04-20 22:59:13 +0200  [28cd229] Generated proper perl module 
2010-04-20 22:57:17 +0200  [0c55614] Corrected RM and 8,141 
2010-04-20 01:04:06 +0200  [30c421f] PerlCritic and others 
2010-04-20 00:38:19 +0200  [16fbdb4] Changed qq to q for all files 
2010-04-19 23:29:31 +0200  [902e27b] Changed a few things in RM handling 
2010-04-19 01:25:18 +0200  [19751f1] Added options, corrected lots of stuff 
2010-04-18 21:50:45 +0200  [2cd1e37] Added CCSDS_Common to move crc there 
2010-04-18 21:20:46 +0200  [d01a5e9] Cleaned a bit RM handling 
2010-04-17 22:59:36 +0200  [ee25210] Loop decoder 
2010-04-17 22:59:00 +0200  [a66ee0e] Added TimePacket, RM Decoding, SW Error Log decoding 
2010-04-03 11:56:54 +0200  [f494931] Changed Crc function 
2010-03-31 22:57:04 +0200  [746cc1b] Removed comments and minimal pretty printer 
2010-03-31 22:56:23 +0200  [4aa9566] Corrected errors on Arrays 
2010-03-31 22:53:49 +0200  [09ae019] RM Struct changed 
2010-03-31 22:53:12 +0200  [3bb4f38] Forgotten the TimeStamp 
2010-03-31 21:48:55 +0200  [a37c842] Added more Pids 
2010-03-31 09:26:24 +0200  [c4089de] Added Debug 
2010-03-31 01:55:30 +0200  [3d49d95] decoder.pl can now read blocks of both types 
2010-03-31 00:49:26 +0200  [436b8dd] test3.pl now read transparently both type of input 
2010-03-30 01:59:32 +0200  [72f3ebe] Changed test2.pl to point to test3.pl 
2010-03-30 01:55:44 +0200  [8187cd1] Added crc cccit check in test3.pl for now 
2010-03-29 19:58:59 +0200  [8feeace] Added better slice handling 
2010-03-29 09:14:22 +0200  [5e82620] Added License at start of each file 
2010-03-29 09:01:28 +0200  [aa1c782] Corrected syntax error 
2010-03-29 08:54:44 +0200  [d278162] To be used as cli tool 

TMPrinter.pm
use strict;
use warnings;
#<<'#';
#    Copyright (C) 2010  Laurent Kislaire, teebeenator@gmail.com
#
#    This program is free software: you can redistribute it and or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#    You should have received a copy of the GNU General Public License
#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
#

package TMPrinter;
use Switch;
use Data::Dumper;

$Data::Dumper::Terse = 1;
$Data::Dumper::Indent = 3;

sub TMPrint {

	my ($tm) = shift;
	
	my $Src_Packet=$tm;
	if (exists $tm->{'TM Source Packet'}) {
	# Then $tm is a scos packet, and we go directly to the TM Packet
		$Src_Packet=$tm->{'TM Source Packet'};
	}
	my $Pkt_Data=$Src_Packet->{'Packet Data Field'}->{'Data Field'};
	my $Pus_Header=$Pkt_Data->{'TMSourceSecondaryHeader'};
	
	my $OBT=${$Pus_Header->{'Sat_Time'}}{'OBT'};
	my $Pus_Type=$Pus_Header->{'Service Type'};
	my $Pus_SubType=$Pus_Header->{'Service Subtype'};
	
	my $Pus_Data=$Pkt_Data->{'PusData'};

	print "TM PUS($Pus_Type,$Pus_SubType)\t";
	printf "OBT=%10.3f\t", $OBT;
	switch ( join(',',$Pus_Type,$Pus_SubType) ){
		case /3,10|3,12/ {
			my $SID=$Pus_Data->{'Sid'};
			my $Col=$Pus_Data->{'Collection Interval'};
			my $Status=$Pus_Data->{'Status'};
			my $NPar=$Pus_Data->{'NPar'};
			my $Params=$Pus_Data->{'ParamId'};
			print "HK/Diag Report Definition  Sid=$SID  ColInt=$Col  Status=$Status  NPar=$NPar\n";
		 	print "Params=\n" . Dumper($Params) if ($NPar);
		}
		case '5,132' {
			my $NEid=$Pus_Data->{'NEid'};
			my $Eids=$Pus_Data->{'Eids'};
			#$Data::Dumper::Indent = 2;
		 	print "Enabled Events  NEids=$NEid\n";
			print "EIDs=\n" . Dumper($Eids) if ($NEid);
		}
		case '6,6' {
			my $MemId=$Pus_Data->{'Memory Id'};
			my $StartAddress=$Pus_Data->{'Start Address'};
			my $Length=$Pus_Data->{'Length'};
			my $dta=$Pus_Data->{'Data'};
			print "Memory Id:$MemId  StartAddress:" . sprintf('%08x',$StartAddress) . "  Length: $Length Data:\n";
      my $l=32;
			for(my $i=0;$i<$Length;$i+=$l){
				my $j=$i+$l-1; $j=$Length-1 if ($j>$Length-1);
				print sprintf('%08x : ',$i). join(' ',	map { sprintf '%02X',$_} @$dta[$i..$j]) . "\n";	
			}
		}
		case '8,140' {
			my $FID=$Pus_Data->{'Function Id'};
			print "Function ID=$FID  ";	
			my $Function_Data=$Pus_Data->{'Data'};
			switch ($FID) {
		 		case 103 {
					my $Ptr=$Function_Data->{'Pointer'};
					my $Entry=$Function_Data->{'RMLogEntry'};
					print "RM Log  Pointer:$Ptr\n";
					my $i=0;
					for(;(my $cTimeStamp=(my $cEntry=$$Entry[$i])->{'TimeStamp'})!=0;$i++) {
            print '-' x 48 ."\n              Entry $i at ". sprintf('%07f',$cTimeStamp)."s          \n".'-' x 48 . "\n";
						my $cSts=$cEntry->{'Status Input'};
						my $cCond=$cEntry->{'Conditioned Alarm'};
						my $cAtmp=$cEntry->{'Attempt index'};
						my $cPat=$cEntry->{'Pattern No'};

						my @mkeys_s=('au', 'auRed', 'tmEnc', 'pmAct', 'wd', 'rm', 'rmRed', 'buttrRed', 
								'pmBit1', 'sdram', 'pmA', 'pmB', 'pmBit0', 'eeprom');
						my $sts_print='';
						foreach my $mkey_s (@mkeys_s) {     # We print either " +XX" or the same number of spaces
              $sts_print.=($cEntry->{$mkey_s})?' +PMBact':' +PMAact',next if ($mkey_s eq 'pmAct');
							$sts_print.=($cEntry->{$mkey_s})?" +$mkey_s":' ' x (2+length($mkey_s));
						}

						my @mkeys_c=( 'WDA', 'WDB', 'EPA', 'EPB', 'BatA', 'BatB', 'Thr', 'SunLoss', 
							    'TempCtl', 'PMAhw', 'PMAall', 'PMAuv', 'PMAsw', 'PMBhw', 'PMBall', 
							    'PMBuv', 'PMBsw', 'SelPM', 'Sep1', 'Sep2', 'Sep3', 'WDen');
						my $cond_print='';
						foreach my $mkey_c (@mkeys_c) {
							$cond_print.=($cEntry->{$mkey_c})?' +SelPMB':' +SelPMA',next if ($mkey_c eq 'SelPM');
							$cond_print.=" +$mkey_c" if $cEntry->{$mkey_c};
						}

						print 'Pattern No        :      '. sprintf('%04d',$cPat) . "\n";
						print 'Status Input      :0x'. sprintf('%08x',$cSts) . " $sts_print\n";
						print 'Conditionned Alarm:  0x'. sprintf('%06x',$cCond) . " $cond_print\n";
						print 'Attempt Index     :  '. sprintf('%8d',$cAtmp) . "\n";
					}
					print '-' x 48 . "\n" if ($i);
				}
				else { print "Undecoded Function\n" . Dumper($Function_Data); }
			}
		}
		case '8,141' {
			my $FID=$Pus_Data->{'Function Id'};
			my $NrCurSlice=$Pus_Data->{'Nr Current Slice'};
			my $TotSlice=$Pus_Data->{'Total Slice'};
			my $Length=$Pus_Data->{'Length'};
			print "Sliced Function Report : Function ID=$FID  Slice:$NrCurSlice/$TotSlice Length of Datas:$Length\n";
			my $dta=$Pus_Data->{'Data'};
			if ($FID==152) {
				print Dumper($dta);
			} else {
			for(my $i=0;$i<$Length;$i+=32){
				my $j=$i+31; $j=$Length-1 if ($j>$Length-1);
				print sprintf('%08x : ',$i). join(' ',	map { sprintf '%02X',$_} @$dta[$i..$j]) . "\n";	
			}
			}
		}
		case '11,19' {
			my $N=$Pus_Data->{'N'};
			print "NParam=$N\n";
			for(my $i=0;$i<$N;$i++) {
				my $cSts=$Pus_Data->{'Status'}->[$i];
				my $PID=$cSts->{'Pidb'}->{'PID'};
				my $Status=$cSts->{'Status'};
				print sprintf('%12s',$PID)."  Status=$Status\n";
			}
		}
		case '12,11' {
			my $N=$Pus_Data->{'N'};
			print "NParam=$N\n";
			for(my $i=0;$i<$N;$i++) {
#TODO				my $cLimit=$$Pus_Data->
				my $ParamId=$Pus_Data->{'ParamId'};
				my $Mask=$Pus_Data->{'Mask'};
				my $ParamValue=$Pus_Data->{'Parameter Value'};
				my $LimCrossed=$Pus_Data->{'Limit crossed'};
				my $PrevCheckSts=$Pus_Data->{'Previous Checking Status'};
				my $CurCheckSts=$Pus_Data->{'Current Checking Status'};
				my $Time=${$Pus_Data->{'Sat_Time'}}{'OBT'};
				print "ParamId=$ParamId  Mask=$Mask  Param Value=$ParamValue  Limit crossed=$LimCrossed  Previous Chk Status=$PrevCheckSts  Current Chk Status=$CurCheckSts  OBT=$OBT\n";
			}
		}
		case '14,4' {
			my $N1=$Pus_Data->{'N1'};
			print "Number of TM Source(N1)=$N1\n";
			for(my $i=0;$i<$N1;$i++) {
				my $cTM=$Pus_Data->{'TMSourcePacket'}->[$i];
				my $PID=$cTM->{'Pidb'}->{'PID'};
				my $FStat=$cTM->{'FStat'};
				my $N2=$cTM->{'N2'};
				print sprintf('%12s',$PID)."  FStat=$FStat  Number of Type(N2)=$N2\n";
				for(my $j=0;$j<$N2;$j++) {
					my $cLType=$cTM->{'Types'}->[$j];
					my $Type=$cLType->{'Type'};
					my $TFStat=$cLType->{'FStat'};
					my $N3=$cLType->{'N3'};
					print "\tType=$Type  FStat=$TFStat  Number of SubType(N3)=$N3\n";
					for(my $k=0;$k<$N3;$k++) {
						my $cLSubType=$cLType->{'SubTypes'}->[$k];
						my $SubType=$cLSubType->{'SubType'};
						my $STFStat=$cLSubType->{'FStat'};
						print "\t\tSubType=$SubType  FStat=$STFStat \n";
					}
				}
			}
		}
		case '14,8' {
			my $N1=$Pus_Data->{'N1'};
			print "Number of Pids(N1)=$N1\n";
			for(my $i=0;$i<$N1;$i++) {
				my $cPid=$Pus_Data->{'Pids'}->[$i];
				my $PID=$cPid->{'Pidb'}->{'PID'};
				my $N2=$cPid->{'N2'};
				print sprintf('%12s',$PID)."  Number of Sids(N2)=$N2\n";
				for(my $j=0;$j<$N2;$j++) {
					my $cSid=$cPid->{'Sids'}->[$j];
					my $SID=$cSid->{'SID'};
					my $FStat=$cSid->{'FStat'};
					print "\t\tSID=$SID  FStat=$FStat\n";
				}
			}
		}
		case '15,6' {
			my $N1=$Pus_Data->{'N1'};
			print "Number of Pids(N1)=$N1\n";
			for(my $i=0;$i<$N1;$i++) {
				my $cTM=$Pus_Data->{'Pids'}->[$i];
				my $PID=$cTM->{'Pidb'}->{'PID'};
				my $StoreId1=$cTM->{'StoreId1'};
				my $N2=$cTM->{'N2'};
				print sprintf('%12s',$PID)."  StoreId1=$StoreId1  Number of Type(N2)=$N2\n";
				for(my $j=0;$j<$N2;$j++) {
					my $cLType=$cTM->{'Types'}->[$j];
					my $Type=$cLType->{'Type'};
					my $StoreId2=$cLType->{'StoreId2'};
					my $N3=$cLType->{'N3'};
					print "\t\tType=$Type  StoreId2=$StoreId2  Number of SubType(N3)=$N3\n";
					for(my $k=0;$k<$N3;$k++) {
						my $cLSubType=$cLType->{'SubTypes'}->[$k];
						my $SubType=$cLSubType->{'SubType'};
						my $StoreId3=$cLSubType->{'StoreId3'};
						print "\t\t\t   SubType=$SubType  StoreId3=$StoreId3 \n";
					}
				}
			}
			
		}
		case '15,13' {
			my $N=$Pus_Data->{'N'};
			print "Number of Stores=$N\n";
			for(my $i=0;$i<$N;$i++) {
				my $cStore=$Pus_Data->{'Stores'}->[$i];
				my $cStoreId=$cStore->{'StoreId'};
				my $cStoreTime1=$cStore->{'Storage Time1'};
				my $cStoreTime2=$cStore->{'Storage Time2'};
				my $cPercent=$cStore->{'Percentage filled'};
				print "StoreId $cStoreId  Time:[${cStoreTime1}s .. ${cStoreTime2}s]  Filled: $cPercent%\n";
			}
		}
		case '19,7' {
			my $N=$Pus_Data->{'N'};
			print "NParam=$N\n";
			$Data::Dumper::Indent = 2;
			for(my $i=0;$i<$N;$i++) {
				my $TC=$Pus_Data->{'TC'}->[$i];
#				my $TC=$$Pus_Data{'TC'}[$i];
				my $PID=$TC->{'Pidb'}->{'PID'};
				my $EID=$TC->{'EID'};
				my $ActSts=$TC->{'Action Status'};
				my $TCType=$TC->{'Service Type'};
				my $TCSubType=$TC->{'Service SubType'};
		 		print "PID=$PID  EID=$EID  Action Status=$ActSts TCPUS($TCType,$TCSubType)  Data:\n" . Dumper($TC->{'TC Application Data'});
			}
		}
		else { print "Undecoded Type/SubType\n" . Dumper($Pus_Data); }
	}
#	print "\n";
}

require Exporter;
our @ISA = qw(Exporter);
our @EXPORT = qw(TMPrint);

1;

